external/external.go

```diff
@@ -136,45 +113,41 @@ func (e *External) CreateService(writer http.ResponseWriter, request *http.Reque
 //
 // The name of the Service comes from the URL the request is sent to and the
 // namespace is a daemon-wide configuration setting.
-func (e *External) UpdateService(writer http.ResponseWriter, request *http.Request) {
-	defer request.Body.Close()
+func (e *External) UpdateService(c echo.Context) error {
 
 	var (
 		service string
-		ok      bool
-		v       = mux.Vars(request)
+		err     error
 	)
 
-	if service, ok = v["name"]; !ok {
-		common.Error(writer, "missing service name in the URL", http.StatusBadRequest)
-		return
+	service = c.Param("name")
+	if service == "" {
+		return &echo.HTTPError{
+			Code:     http.StatusBadRequest,
+			Internal: errors.New("missing service name in the URL"),
+		}
 	}
 
 	log.Printf("UpdateService: updating service %s", service)
 
-	buf, err := ioutil.ReadAll(request.Body)
-	if err != nil {
-		common.Error(writer, err.Error(), http.StatusInternalServerError)
-		return
-	}
-
 	opts := &ServiceOptions{}
-
-	err = json.Unmarshal(buf, opts)
-	if err != nil {
-		common.Error(writer, err.Error(), http.StatusBadRequest)
-		return
+	if err = c.Bind(opts); err != nil {
+		return err
 	}
 
 	if opts.TargetPort == 0 {
-		common.Error(writer, "TargetPort was either not set or set to 0", http.StatusBadRequest)
-		return
+		return &echo.HTTPError{
+			Code:     http.StatusBadRequest,
+			Internal: errors.New("TargetPort was either not set or set to 0"),
+		}
 	}
 	log.Printf("UpdateService: target port for %s should be %d", service, opts.TargetPort)
 
 	if opts.ListenPort == 0 {
-		common.Error(writer, "ListenPort was either not set or set to 0", http.StatusBadRequest)
-		return
+		return &echo.HTTPError{
+			Code:     http.StatusBadRequest,
+			Internal: errors.New("ListenPort was either not set or set to 0"),
+		}
```

Comment:
This is an improvement for later, but it might be nice to have common validations such as these in a single validation function.

Reply:
Yeah, one of my todo's in the near future is look into the validation library you recommended a while back and start integrating it.